### ISSUE TITLE
problems with backing up individual files

### ISSUE THREAD

**Issue by OCat49** [cccapon/KeepBack#3]
Chris,

Sorry to bother you again.

Up to now, I've been archiving a whole directory ("folder") and its contents.  Now, I've decided I need to include a couple of files deep under another folder.  Obviously, I don't understand this.

I tried adding just those two files.  It archived EVERYTHING under the parent folder.

I went and read the doc you provide again.  I thought I'd figured out what I needed.  I put in "*" as an exclude filter, then put in the full paths (relative to the parent folder) of the two files.  It backed the two files up, but also included all folders under the parent (as empty folders).

At this point, I started doing some tests instead of the real control file.  I've tried a bunch of variations, but nothing works.  Mostly, they consisted of trying adding "*/" as another exclude filter.  And the ONLY thing that gets backed up is the .keep control file.  My last try was to add in all the intermediate levels of folders.  No luck.

What I need is the two files, with their paths, but not all the other files and/or folders.

I'll attach my test .keep file.  I'm sure there's a way to do what I want.  I've just been unable to figure it out.  Please tell me what I'm doing wrong.

Again, sorry to bother you!

And thank you.

Hmm.  It won't accept a .keep file.  I'll just paste in the contents here.  Or no, I'll zip it up & attach the zip.  If you want earlier versions, I believe I can supply them.

-- Dave

[kbtest.zip](https://github.com/cccapon/KeepBack/files/5578863/kbtest.zip)

**Comment by OCat49** [comment]
Chris,
One other thing I didn't notice earlier.  For the actual backup, I had "*" as the only exclusion.  As I said, it backed up all the empty folders.  This time, I looked at the log.  There were lots of protection violations even in trying to do that much.

Last night when I first ran into this problem, I guess I didn't really look into the log in detail, so I didn't see that.

Thanks.

-- Dave

**Comment by OCat49** [comment]
Chris,

This says I closed it.  If I did, I didn't mean to and wasn't aware of it.  I'm reopening.  I still haven't found a way around this.

Thanks.

**Comment by cccapon** [comment]
It is possible to do what you want, but it is ugly.  I will try to improve this in the next release.

1. The order of filters is important.  They are applied top to bottom with the last match taking effect.
2. Create Exclude filters for "*" and "*/"  -- this excludes all files and all folders.
3. Create Include filters for the root folder "/" and then for each folder level down to where your files are stored.
4. Create Include filters for each file.

Here is an example:
Including only two files (Process.ctrl and Run-Report.ctrl) along with their sub folders...

          Exclude  *
          Exclude  */
          Include  /
          Include  /case/
          Include  /case/0002/
          Include  /case/0002/Ctrl/
          Include  /case/0002/Ctrl/Process.ctrl
          Include  /case/0002/Ctrl/Run-Report.ctrl

Cheers.

**Comment by OCat49** [comment]
Chris,

Thank you for your very prompt reply.

I'd figured out the top-to-bottom order.  I think I understand your
procedure.  I was trying something a little like this, but not exactly
right.  I'd missed the step of including each & every level.  I hope I've
got it now.

In your steps 2 & 3 (comparing them to your example), it looks like
asterisks inside the quotation marks are missing.  I hope that's correct.
I'll try it tonight, anyway.  Sigh.

--
-- Dave Lovelace
   davelovelace at gmail dot com
"English doesn't just borrow from other languages.  English follows other
languages down dark alleys, knocks them over, and goes through their
pockets for loose grammar." -- James Nicoll

**Comment by cccapon** [comment]
The quotation marks in the explanation are not included in the filters.  So, follow the example text but substitue your own file paths.

In the next release, I will try to have the Include parent folders added automatically.

**Comment by OCat49** [comment]
Chris,
Thank you.  Your example did the trick.  I had the right idea, but not the details.

After that first backup, I still had all those empty folders.  But as I realized, that was because they were carried over from the previous backup.  Before the next backup, I went in and nuked them all (a bit of a job, since I wanted to keep the two files and their parents).  That probably wasn't the correct way to do it, but it did the trick.

As for the quotes:  I understood the quotes.  My point was that the asterisks that should have been inside them weren't there.  They were there in your example, though.

Again, thank you!
